File: 6-1-1.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        int[][] A = new int[5][5];
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                A[i][j] = scan.nextInt();
            }
        }
        int sr = scan.nextInt();
        int sc = scan.nextInt();
        int tr = 0, tc = 0;
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                if (A[i][j] == 1) {
                    tr = i;
                    tc = j;
                }
            }
        }
        int[] dr = { -1, 1, 0, 0 };
        int[] dc = { 0, 0, -1, 1 };
        int[][] visited = new int[5][5];
        int[][] dist = new int[5][5];
        Deque<Pair<Integer, Integer>> Q = new ArrayDeque<>();
        Q.add(new Pair<>(sr, sc));
        visited[sr][sc] = 1;
        dist[sr][sc] = 0;
        while (!Q.isEmpty()) {
            Pair<Integer, Integer> now = Q.pollFirst();
            int r = now.getFirst();
            int c = now.getSecond();
            if (r == tr && c == tc) {
                System.out.println(dist[r][c]);
                System.exit(0);
            }
            for (int i = 0; i < 4; i++) {
                int nr = r + dr[i];
                int nc = c + dc[i];
                if (inRange(nr, nc) && visited[nr][nc] == 0 && A[nr][nc] != -1) {
                    Q.add(new Pair<>(nr, nc));
                    dist[nr][nc] = dist[r][c] + 1;
                    visited[nr][nc] = 1;
                }
            }
        }
        System.out.println(-1);
    }

    public static boolean inRange(int r, int c) {
        return 0 <= r && r <= 4 && 0 <= c && c <= 4;
    }
}

class Pair<T, U> {
    private T first;
    private U second;

    public Pair(T first, U second) {
        this.first = first;
        this.second = second;
    }

    public T getFirst() {
        return first;
    }

    public U getSecond() {
        return second;
    }
}

File: 6-1-1.cpp
#include<bits/stdc++.h>
using namespace std;
typedef pair<int, int> pii;
int in_range(int r, int c) {
    return 0 <= r && r <= 4 && 0 <= c && c <= 4;
}
int main(int argc, char* argv[]) {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int A[5][5];
    for (int i = 0; i < 5; i++) {
        for (int j = 0; j < 5; j++) {
            cin >> A[i][j];
        }
    }
    int sr, sc; cin >> sr >> sc;
    int tr, tc;
    for (int i = 0; i < 5; i++) {
        for (int j = 0; j < 5; j++) {
            if (A[i][j] == 1) {
                tr = i; tc = j;
            }
        }
    }

    int dr[4] = { -1, 1, 0, 0 }, dc[4] = { 0, 0, -1, 1 };
    int visited[5][5] = { 0 }, dist[5][5] = { 0 };
    deque<pii> Q;
    Q.push_back({ sr, sc });
    visited[sr][sc] = 1;
    dist[sr][sc] = 0;
    while (!Q.empty()) {
        pii now = Q.front(); Q.pop_front();
        int r = now.first, c = now.second;
        if (r == tr && c == tc) {
            cout << dist[r][c]; exit(0);
        }
        for (int i = 0; i < 4; i++) {
            int nr = r + dr[i], nc = c + dc[i];
            if (in_range(nr, nc) && visited[nr][nc] == 0 && A[nr][nc] != -1) {
                Q.push_back({ nr,nc });
                dist[nr][nc] = dist[r][c] + 1;
                visited[nr][nc] = 1;
            }
        }
    }
    cout << -1;
}